Platform/AMD/AmdPlatformPkg/DynamicTables/Library/Acpi/AcpiSsdtPciLib/AcpiSsdtPciLib.c

```diff
@@ -0,0 +1,1540 @@
+                 PrtNode
+                 );
+      if (EFI_ERROR (Status)) {
+        ASSERT_EFI_ERROR (Status);
+      }
+    }
+
+    // Attach the _PRT entry.
+    Status = AmlAttachNode (DeviceNode, PrtNode);
+    if (EFI_ERROR (Status)) {
+      AmlDeleteTree (PrtNode);
+      ASSERT_EFI_ERROR (Status);
+    }
+
+    PrtNode = NULL;
+  }
+
+  if (EFI_ERROR (Status)) {
+    Status = EFI_DEVICE_ERROR;
+  }
+
+  return Status;
+}
+
+/**
+  Insert CXL Root Bridge into the AML table
+
+  @param[in]      RootBridgeHead  - RootBridge information pointer
+  @param[in]      RootBridgeCount - Number of root bridges
+  @param[in,out]  PciNode         - AmlLib table node
+
+  @retval         EFI_SUCCESS, various EFI FAILUREs.
+**/
+EFI_STATUS
+EFIAPI
+InternalInsertCxlRootBridge (
+  IN          AMD_PCI_ROOT_BRIDGE_OBJECT_INSTANCE  *RootBridgeHead,
+  IN          UINTN                                RootBridgeCount,
+  IN OUT      AML_OBJECT_NODE_HANDLE               PciNode
+  )
+{
+  AMD_PCI_ADDR                         PciAddr;
+  AMD_PCI_ROOT_BRIDGE_OBJECT_INSTANCE  *RootBridge;
+  AML_METHOD_PARAM                     MethodParam[7];
+  AML_OBJECT_NODE_HANDLE               CrsNode;
+  AML_OBJECT_NODE_HANDLE               DeviceNode;
+  AML_OBJECT_NODE_HANDLE               OscMethod;
+  AML_OBJECT_NODE_HANDLE               PackageNode;
+  CHAR8                                NameSeg[5];
+  EFI_STATUS                           Status;
+  UINT32                               EisaId;
+  UINT8                                DevIndex;
+  UINT8                                Index;
+
+  DEBUG ((DEBUG_INFO, "%a: Entry\n", __func__));
+  ZeroMem ((VOID *)&PciAddr, sizeof (PciAddr));
+
+  //
+  // Populate the data structure for the CXL devices in the system to add to
+  // the ACPI Table
+  //
+  DevIndex = 0;
+  for (Index = 0, RootBridge = RootBridgeHead; Index < RootBridgeCount; Index++, RootBridge++) {
+    if ((RootBridge->CxlCount == 0) || (RootBridge->CxlPortInfo.IsCxl2 == TRUE)) {
+      continue;
+    }
+
+    DevIndex++;
+
+    CopyMem (NameSeg, "CXLx", AML_NAME_SEG_SIZE + 1);
+    if (DevIndex < 0x10) {
+      NameSeg[AML_NAME_SEG_SIZE - 1] = AsciiFromHex (DevIndex);
+    } else {
+      NameSeg[AML_NAME_SEG_SIZE - 2] = AsciiFromHex (DevIndex);
+    }
+
+    Status = AmlCodeGenDevice (NameSeg, PciNode, &DeviceNode); // RootBridge
+    if (EFI_ERROR (Status)) {
+      ASSERT_EFI_ERROR (Status);
+      return Status;
+    }
+
+    Status = AmlCodeGenNameString ("_HID", "ACPI0016", DeviceNode, NULL);
+    if (EFI_ERROR (Status)) {
+      ASSERT_EFI_ERROR (Status);
+      return Status;
+    }
+
+    Status = AmlCodeGenNamePackage ("_CID", DeviceNode, &PackageNode);
+    ASSERT_EFI_ERROR (Status);
+
+    // Name (_CID, EISAID("PNP0A03"))
+    Status = AmlGetEisaIdFromString ("PNP0A03", &EisaId);
+    if (EFI_ERROR (Status)) {
+      ASSERT_EFI_ERROR (Status);

```

Comment:
Same here, do we have to delete PackageNode in the fail conditions?

Reply:
done, added the exit_handler